File: tests/fixtures/ldap_injection/python/safe_ldap_search.py
# Safe: user-supplied substring run through `escape_filter_chars` (RFC 4515)
# before being concatenated into the filter.  The sanitizer clears the
# LDAP_INJECTION cap so the sink does not fire.
import ldap
from ldap.filter import escape_filter_chars
from flask import request


def lookup():
    conn = ldap.initialize("ldap://example.com")
    user = request.form["user"]
    safe = escape_filter_chars(user)
    flt = "(uid=" + safe + ")"
    return conn.search_s("ou=people,dc=example,dc=com", ldap.SCOPE_SUBTREE, flt)

File: tests/fixtures/ldap_injection/java/SafeLdapSearch.java
// Safe: the user-supplied substring is run through Spring LDAP's
// LdapEncoder.filterEncode (RFC 4515 escape) before being assembled into the
// filter.  The Sanitizer(LDAP_INJECTION) clears the cap and the sink does not
// fire.
import javax.naming.directory.DirContext;
import javax.naming.directory.SearchControls;
import javax.servlet.http.HttpServletRequest;
import org.springframework.ldap.support.LdapEncoder;

public class SafeLdapSearch {
    private DirContext ctx;

    public Object lookup(HttpServletRequest req) throws Exception {
        String user = req.getParameter("user");
        String safe = LdapEncoder.filterEncode(user);
        String filter = "(uid=" + safe + ")";
        return ctx.search("ou=people,dc=example,dc=com", filter, new SearchControls());
    }
}
